bin/knapsack_pro_all.rb

```diff
@@ -1,176 +1,174 @@
 COMMANDS = %{
-./bin/knapsack_pro_a_few_commands 0 2
-./bin/knapsack_pro_a_few_commands 1 2
+./bin/knapsack_pro_a_few_commands 0 2 some-build-id
```

Comment:
The downside of using constant `some-build-id` as CI build ID is that you would run tests only once. When you run `bin/knapsack_pro_all.rb` for the second time (then for Queue Mode cases) you would run no tests. So there is no way to verify if knapsack pro breaks something during tests runtime.

I would generate a random ID as CI build for each command. Please remember CI build ID must be the same for node index 0 and 1 so that the commands consume from the same Queue. The bin scripts run in sync way so the node index 1 would run no tests because node index 0 consumed the queue. That's fine. This way we test both scenarios: consuming from queue and when queue is already consumed.

```
require 'securerandom'

ci_build_id = SecureRandom.uuid # "8fedd1c9-306e-46e7-a4f5-ff5336943c9b"

then run:

./bin/knapsack_pro_a_few_commands 0 2 #{ci_build_id}
./bin/knapsack_pro_a_few_commands 1 2 #{ci_build_id}

For another command you need to generate another unique CI build ID:

```
another_ci_build_id = SecureRandom.uuid

./bin/knapsack_pro_rspec 0 2 #{another_ci_build_id}
./bin/knapsack_pro_rspec 1 2 #{another_ci_build_id}
```

You would have to refactor this code somehow to be able to generate a CI build ID and share it per bin script.

Reply:
Makes sense. I pushed a change to take care of this. WDYT?

I'm wondering if we shouldn't move these tests to RSpec and then parallelize them with KS. But I'm not completely sure I understand the context of why they exist in bash (bin/) in the first place.

Could you share more details on how/when you use them?

Reply:
> Makes sense. I pushed a change to take care of this. WDYT?

I like it. Thank you.

> I'm wondering if we shouldn't move these tests to RSpec and then parallelize them with KS. But I'm not completely sure I understand the context of why they exist in bash (bin/) in the first place.
> 
> Could you share more details on how/when you use them?


I use these bin scripts mainly for development purposes when I need to test a new feature in the knapsack_pro gem. Or when I want to record a new CI build on the API side so that I can test something in the user dashboard in development. 

I'm using the bin scripts locally as an integration test to see if things work as expected:

knapsack_pro gem + example Rails test suites + API = all 3 works corrrectly.

After I develop a new feature for API or for the knapsack_pro gem I run all bin scripts locally as a final check to ensure nothing is accidentally broken.


> I'm wondering if we shouldn't move these tests to RSpec and then parallelize them with KS.

No need. API is covered by RSpec tests. Gem is covered by RSpec tests.
Bin scripts are just integration tests to ensure things work when connected together.

On CircleCI we run some tests `.circleci/config.yml` but not all bin scripts because it would be slow.

Bin scripts are just to see how knapsack_pro gem would beave in all possible configuration scenarios. It's also helpful when I would change something on the API side and I'd like to test what happens in API responses, I can see it in the knapsack_pro gem logs. I see how the knapsack_pro gem would act (for example what happens when API is down, or when I simulate a timeout from API etc).